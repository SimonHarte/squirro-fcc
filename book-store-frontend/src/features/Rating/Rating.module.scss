@import "../../styles/variables.scss";

.wrapper {
  display: flex;
  flex-direction: row-reverse;
  gap: $spacing-base;
}

.star {
  &_current {
    ~ .label {
      color: orange;
    }
  }
  &:focus {
    ~ .label {
      color: gold;
    }
  }
}

.label {
  color: gray;
  cursor: pointer;

  &::before {
    content: "â˜…";
    display: block;
    font-size: 1.5rem;
    line-height: 1ch;
  }

  &:hover {
    color: gold;

    ~ .label {
      color: gold;
    }
  }
}
