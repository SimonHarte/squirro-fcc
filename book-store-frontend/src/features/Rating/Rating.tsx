import { Fragment, type ReactElement, type FC, type ChangeEvent } from 'react';
import classNames from 'classnames';
import { useSetNewRatingMutation } from '../../app/api';
import type { RatingProps } from './Rating.types';
import styles from './Rating.module.scss';

export const RATING_RANGE = {
  min: 1,
  max: 5,
};

const Rating: FC<RatingProps> = ({
  storeName,
  currentRating,
}: RatingProps): ReactElement => {
  const [setNewRating] = useSetNewRatingMutation();

  return (
    <fieldset>
      <legend className={styles.vh}>Rating ({currentRating})</legend>
      <div className={styles.stars}>
        {[...Array(RATING_RANGE.max)].map((_, i) => {
          const groupName = `${storeName}-rating`;
          const id = `${groupName}${i}`;
          const isSelected = currentRating === i + RATING_RANGE.min;
          const rating = RATING_RANGE.max - i;

          return (
            <Fragment key={id}>
              <input
                className={classNames(
                  styles.vh,
                  styles.star,
                  isSelected && styles.star_current
                )}
                id={id}
                name={groupName}
                type="radio"
                value={rating}
                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                  setNewRating({
                    id: i,
                    rating: e.target.value,
                  });
                }}
              />
              <label htmlFor={id} className={styles.label}>
                <span className={styles.vh}>{rating}</span>
              </label>
            </Fragment>
          );
        })}
      </div>
    </fieldset>
  );
};

export default Rating;
