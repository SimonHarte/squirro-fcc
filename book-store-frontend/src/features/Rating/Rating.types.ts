import { ChangeEventHandler } from 'react';

export type RatingProps = {
  storeName: string;
  currentRating: number;
  onAddRating: ChangeEventHandler;
};
