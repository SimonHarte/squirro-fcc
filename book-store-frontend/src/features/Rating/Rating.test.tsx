import { shallow } from 'enzyme';
import Rating from './index';

jest.mock('../../app/api', () => ({
  useSetNewRatingMutation: () => [() => {}],
}));

describe('Rating', () => {
  test('should render correctly', () => {
    const wrapper = shallow(
      <Rating storeName="Test Store" currentRating={2} />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
