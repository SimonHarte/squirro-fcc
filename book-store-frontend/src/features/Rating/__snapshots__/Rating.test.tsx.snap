// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rating should render correctly 1`] = `
<fieldset>
  <legend
    className="vh"
  >
    Rating (
    2
    )
  </legend>
  <div
    className="stars"
  >
    <input
      className="vh star"
      id="Test Store-rating0"
      key="Test Store-rating0"
      name="Test Store-rating"
      onChange={[MockFunction]}
      type="radio"
      value={5}
    />
    <label
      className="label"
      htmlFor="Test Store-rating0"
    >
      <span
        className="vh"
      >
        5
      </span>
    </label>
    <input
      className="vh star star_current"
      id="Test Store-rating1"
      key="Test Store-rating1"
      name="Test Store-rating"
      onChange={[MockFunction]}
      type="radio"
      value={4}
    />
    <label
      className="label"
      htmlFor="Test Store-rating1"
    >
      <span
        className="vh"
      >
        4
      </span>
    </label>
    <input
      className="vh star"
      id="Test Store-rating2"
      key="Test Store-rating2"
      name="Test Store-rating"
      onChange={[MockFunction]}
      type="radio"
      value={3}
    />
    <label
      className="label"
      htmlFor="Test Store-rating2"
    >
      <span
        className="vh"
      >
        3
      </span>
    </label>
    <input
      className="vh star"
      id="Test Store-rating3"
      key="Test Store-rating3"
      name="Test Store-rating"
      onChange={[MockFunction]}
      type="radio"
      value={2}
    />
    <label
      className="label"
      htmlFor="Test Store-rating3"
    >
      <span
        className="vh"
      >
        2
      </span>
    </label>
    <input
      className="vh star"
      id="Test Store-rating4"
      key="Test Store-rating4"
      name="Test Store-rating"
      onChange={[MockFunction]}
      type="radio"
      value={1}
    />
    <label
      className="label"
      htmlFor="Test Store-rating4"
    >
      <span
        className="vh"
      >
        1
      </span>
    </label>
  </div>
</fieldset>
`;
